### PR TITLE
Add Deep Q-Network (DQN) Algorithm to Baselines

### PR DESCRIPTION
**Pull request by affan002**
#### **Overview**
This PR introduces a baseline DQN implementation integrated into the existing multi-agent RL framework. It learns neural network-based policies through experience replay and target networks, complementing the existing tabular IQL baseline.

#### **Core Implementation**
* **`baselines/DQN/dqn.py`**: Main algorithm class inheriting `BaseAlgorithm`. Implements an independent learner per agent with a shared replay buffer and a full training loop (includes checkpointing and TensorBoard/wandb logging).
* **`baselines/DQN/dqn_agent.py`**: Single-agent learner wrapper. Features per-agent online/target networks, Adam optimizer (with gradient clipping), epsilon-greedy action selection, and hard target network synchronization.
* **`baselines/DQN/dqn_network.py`**: Neural network architectures (`DuelingMLP`). Configurable hidden layers and supports both vanilla and dueling architectures.
* **`baselines/DQN/replay_buffer.py`**: Fixed-capacity FIFO buffer with numpy storage and efficient PyTorch tensor conversion for deterministic sampling.
* **`baselines/DQN/utils.py`**: Experiment utilities for timestamped folder creation, checkpointing, and README generation (mirrors IQL utils).

#### **Algorithm Variants Supported**
Controlled via config flags in `configs/dqn_experiment.yaml`:
* **Vanilla DQN**: Single Q-head.
* **Double DQN** (`use_double: true`): Online network selects best action, target evaluates (reduces overestimation bias).
* **Dueling DQN** (`use_dueling: true`): Decomposes into value and advantage streams for better state value learning.

#### **Current State Encoding vs. Future Abstraction**
* **Current**: Includes a fixed-size observation encoder (`_obs_to_vector`) that flattens structured dict observations into 24-element float32 vectors.
* **Proposed Future Architecture**: Move observation encoding to an abstract method `_obs_to_vec()` in `multi_agent_package/observations/base.py`. This will decouple the state dimension from DQN, allowing multiple algorithms to reuse the same encoding across any observation type.

#### **Usage Examples**

**Train via config:**
```bash
cd src
python -m multi_agent_package.scripts.run_from_config
```

**Evaluate trained model:**
```bash
cd src
python scripts/test_dqn.py -checkpoint checkpoints/dqn_checkpoint.pt -episodes 5
```

#### **Files Changed**
```text
src/
├── baselines/
│   ├── __init__.py                    (+ DQN import)
│   └── DQN/
│       ├── __init__.py                (new)
│       ├── dqn.py                     (new)
│       ├── dqn_agent.py               (new)
│       ├── dqn_network.py             (new)
│       ├── replay_buffer.py           (new)
         ├── test_dqn.py                    (new)
│       └── utils.py                   (new)
├── multi_agent_package/scripts/
│   ├── run_from_config.py            
└── configs/
    └── dqn_experiment.yaml            (new)
```

#### **Next Steps**
1. Add `_obs_to_vec()` method to the observation base class for algorithm-agnostic encoding.
2. Implement Prioritized Experience Replay (PER) for faster learning.

### CODE REVIEW

**Comment by ProValarous** [comment]
This PR is depreciated. Thank you for this work.
